Fixes #33909 - Drop data migrations before Foreman 2.0

tbrisker opened the pull request:
All data migrations only modify data on upgrades. On new installs they
are not required. Since we do not generally support upgrading multiple
releases, we can clean up old data migrations. This commit cleans up all
the data migrations that were present before Foreman 2.0.

Second commit, since I know this topic may be controversial:

Refs #33909 - Delete empty migrations

New installs will not run these empty migrations, and old installs will
only list that they have been executed when checking the `rake
db:migrate:status` with no file, which is harmless.
<!---

Thank you for contributing to The Foreman, please read the
[following guide](https://www.theforeman.org/contribute.html), in short:

* [Create an issue](https://projects.theforeman.org/projects/foreman/issues)
* Reference the issue via `Fixes #1234` in the commit message
* Prefer present-tense, imperative-style commit messages
* Mark all strings for translation, see [1]
* Suggest prerequisites for testing and testing scenarios following example above.
* Prepend `[WIP]` for work in progress to prevent bots from triggering actions
* Be patient, we will do our best to take a look as soon as we can
* Explain the purpose of the PR, attach screenshots if applicable
* List all prerequisites for testing (e.g. VMware cluster, two smart proxies...)
* Reviewers often use extensive list of items to check, have a look prior submitting [2]
* Be nice and respectful

1: https://projects.theforeman.org/projects/foreman/wiki/Translating#Translating-for-developers
2: https://github.com/theforeman/foreman/blob/develop/developer_docs/pr_review.asciidoc
-->

tbrisker committed:
Fixes #33909 - Drop data migrations before Foreman 2.0

All data migrations only modify data on upgrades. On new installs they
are not required. Since we do not generally support upgrading multiple
releases, we can clean up old data migrations. This commit cleans up all
the data migrations that were present before Foreman 2.0.

tbrisker committed:
Refs #33909 - Delete empty migrations

New installs will not run these empty migrations, and old installs will
only list that they have been executed when checking the `rake
db:migrate:status` with no file, which is harmless.

ezr-ondrej commented:
Could we add some purge task to remove the migrations from schema_migrations? It would be my only concern with deleting migrations - having a way to remove them from DB (and also having some list of such)

tbrisker committed:
Refs #33909 - Remove useless User lookup

There is no need to lookup the admin user and use it to update the
http proxy fields.

ezr-ondrej commented on db/seeds.d/110-smart_proxy_features.rb:
```diff
@@ -1,5 +1,5 @@
 # Proxy features
-proxy_features = ["TFTP", "DNS", "DHCP", "Puppet CA", "BMC", "Realm", "Facts", "Logs", "HTTPBoot", "External IPAM",
+proxy_features = ["Templates", "TFTP", "DNS", "DHCP", "Puppet CA", "BMC", "Realm", "Facts", "Logs", "HTTPBoot", "External IPAM",
```
Is this related?

tbrisker replied:
yes, It was added in a migration previously instead of the seed.

tbrisker commented:
@ezr-ondrej we could, but I don't see much benefit in doing so - if we keep them, we can tell that they ran in the past even if they no longer exist (i.e. identify if the db is new or was upgraded). 
Having timestamps in schema_migrations without files is valid. Many rails apps delete all old migrations, when the `schema.rb` or `structure.sql` file is committed in source control and is used as the base db structure - which unfortunately we cannot do due to $reasons.

ekohl commented on db/migrate/20100310080727_add_family_to_os.rb:
```diff
@@ -1,29 +1,6 @@
-    Operatingsystem.all.each do |os|
-      case os.name
-      when /RedHat|Centos|Fedora/i
-        os.family_id = Operatingsystem::FAMILIES.index :RedHat
```
Fun fact: I can't see how this worked. AFAIK Families never had an entry which accepted symbols and it always was `Redhat`, not `RedHat`.

ekohl commented on db/migrate/20121118120028_migrate_hypervisors_to_compute_resources.rb:
```diff
@@ -1,18 +1,6 @@
+    drop_table :hypervisors, if_exists: true
   end
 
   def down
```
It's weird to see an `up` migration that's a `drop_table` but an empty `down`.

tbrisker replied:
That's true, although it was that way before as well so no idea what should be in that table. Also, I highly doubt anyone would rollback migrations all they way to 2012 😜

ekohl commented on db/migrate/20140902102858_convert_ipam_to_string.rb:
```diff
@@ -1,30 +1,14 @@
-  end
-
   def up
     add_column :subnets, :ipam_tmp, :string, :default => IPAM::MODES[:dhcp], :null => false, :limit => 255
```
If you're dropping the migration, there's no reason for a tmp column. Can't this be squashed down to a `change_column`? Same comment for `down`.

ekohl commented on db/migrate/20190226182746_create_registration_facets.rb:
```diff
@@ -1,5 +1,3 @@
-# frozen_string_literal: true
-
```
I'm not sure if this is needed.

tbrisker replied:
I don't know why it was added in the first place, but there are no strings in this file so it had no affect anyways.

ekohl commented on db/migrate/20210114143800_nilify_empty_proxy_credentials.rb:
```diff
@@ -1,19 +1,11 @@
 class NilifyEmptyProxyCredentials < ActiveRecord::Migration[6.0]
   def up
-    if User.unscoped.find_by_login(User::ANONYMOUS_ADMIN).present?
-      User.as_anonymous_admin do
```
Was this here to disable auditing or something? This is fairly recent (d60d1cb169ac83a8aed6ba7256c1b731dce07350) so maybe @ares remembers.

tbrisker replied:
I'm not sure, possibly an attempt to bypass taxonomy or authorization scope. `update_all` triggers sql `UPDATE` query directly without initializing AR objects so it wouldn't have been audited in any case.

ekohl reviewed (commented):
I still would have preferred to really squash the migrations so avoid changing tables, but I can live with this as a start.

tbrisker committed:
Refs #33909 - Squash imap column conversion

ekohl commented:
Tests fail with:
```
NoMethodError: undefined method `find_by_lower_login' for #<Class:0x0000000009c55fb8>
Did you mean?  find_by_login
```

tbrisker committed:
Refs #33909 - Reset User column information

While migrations shouldn't rely on any users existing in the DB, there
are some that currently do. If they try to use User.find_by_lower_login
they will currently error out since the method won't be defined yet.

tbrisker commented:
I've re-added the column info reset to the migration that adds lower_login to users. While the specific migration that fails should be fixed by GH-8918, there could potentially be other migrations in plugins that try looking up users (which they shouldn't, but that's outside the scope of this PR to fix).

tbrisker committed:
Refs #33909 - Drop test for deduplicate_subnets module

ezr-ondrej reviewed (approved):
I agree that ultimate goal should be migration squash, but this is definitelly a step in the right direction. So big :+1:

ekohl reviewed (approved):
I wonder if this has a (positive) performance impact or if it's negligible. I'll see this as a step in the right direction to actually squashing.

ezr-ondrej commented:
Thanks @tbrisker ! :)
